netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +100,258 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+      /**
+       * Swallowing any flushes is not only an optimization but also required
+       * for the SslHandler to work correctly. If the SslHandler receives multiple
+       * flushes while the handshake is still ongoing, then the handshake "randomly"
+       * times out. Not sure at this point why this is happening. Doing a single flush
+       * seems to work but multiple flushes don't ...
+       */
+      if (bufferedWrites == null) {
+        ctx.flush();
+      } else {
+        flushRequested = true;
       }
+    }
 
-      @Override
-      public void onConnected(Channel channel) {
-        // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
-        // which causes the upgrade headers to be added
-        DefaultHttpRequest upgradeTrigger =
-            new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
-        channel.writeAndFlush(upgradeTrigger);
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Exception {
+      failBufferedAndClose(ctx);
+    }
+
+    protected void failBufferedAndClose(ChannelHandlerContext ctx) {
+      if (bufferedWrites != null) {
+        Exception e = new Exception("Buffered write failed.");
+        while (!bufferedWrites.isEmpty()) {
+          ChannelWrite write = bufferedWrites.poll();
+          write.promise.setFailure(e);
+        }
+        bufferedWrites = null;
       }
-    };
-  }
+      /**
+       * In case something goes wrong ensure that the channel gets closed as the
+       * NettyClientTransport relies on the channel's close future to get completed.
+       */
+      ctx.close();
+    }
 
-  /**
-   * Create a "no-op" negotiation that simply assumes the protocol to already be negotiated.
-   */
-  public static Negotiation plaintext(final ChannelHandler... handlers) {
-    final ChannelInitializer<Channel> initializer = new ChannelInitializer<Channel>() {
-      @Override
-      public void initChannel(Channel ch) throws Exception {
-        ch.pipeline().addLast(handlers);
+    protected void writeBufferedAndRemove(ChannelHandlerContext ctx) {
+      if (!ctx.channel().isActive() || writing) {
+        return;
       }
-    };
-    return new Negotiation() {
-      private final SettableFuture<Void> completeFuture = SettableFuture.create();
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+      // Make sure that method can't be reentered, so that the ordering
+      // in the queue can't be messed up.
+      writing = true;
+      while (!bufferedWrites.isEmpty()) {
+        ChannelWrite write = bufferedWrites.poll();
+        ctx.write(write.msg, write.promise);
       }
-
-      @Override
-      public void onConnected(Channel channel) {
-        completeFuture.set(null);
+      if (flushRequested) {
+        ctx.flush();
```

Comment:
Although it really _shouldn't_ matter, you may want to move the bufferedWrites = null to before flush(), since flush() could queue more writes.